[SE-0253] Introduce callables.

dan-zheng opened the pull request:
Implementation for [SE-0253](https://github.com/apple/swift-evolution/blob/master/proposals/0253-callable.md).

Introduce callables: values that can be called like functions.
Methods named `func callFunction` act as call-syntax delegate methods.

```swift
struct Adder {
  var base: Int
  func callAsFunction(_ x: Int) -> Int {
    return x + base
  }
}
var adder = Adder(base: 3)
adder(10) // desugars to `adder.callAsFunction(10)`
```

dan-zheng committed:
Introduce callables.

Introduce callables: values that can be called like functions.
Methods named `func callFunction` act as call-syntax delegate methods.

```
struct Adder {
  var base: Int
  func callFunction(_ x: Int) -> Int {
    return x + base
  }
}
var adder = Adder(base: 3)
adder(10) // desugars to `adder.callFunction(10)`
```

dan-zheng commented on include/swift/AST/Decl.h:
```diff
@@ -5812,6 +5812,9 @@ class FuncDecl : public AbstractFunctionDecl {
     return getSelfAccessKind() == SelfAccessKind::__Consuming;
   }
+  bool isCallable() const {
+    return getName().str() == "call" && isInstanceMember();
```
Consider adding an `IsCallable` bit to `FuncDecl`?

@lattner's [response](https://github.com/apple/swift/pull/24122#discussion_r277122262):

> I'd recommend _not_ doing that, unless this is a particularly hot predicate. Cached bits like this have to be kept in sync with the ast. Also, instead of serializing it, if the bit needs to be stored, it can always be computed in the module reader.

dan-zheng commented on lib/Sema/CSApply.cpp:
```diff
@@ -7323,6 +7326,59 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+    // FIXME(TF-444): This logic for `mutating` method calls incorrectly rejects
+    // IUOs. Performing this ad-hoc check using `originalFn` feels hacky;
+    // rewrite if possible.
+    if (!cs.getType(originalFn)->hasLValueType() && selfParam->isInOut()) {
```
Todo: find a better way to handle/diagnose `mutating func call` methods.
Then, re-enable `mutating func call` + IUO test.

dan-zheng replied:
Constraint system experts: do you have advice on supporting `mutating func callFunction` methods more robustly? Perhaps appropriate constraints can be generated in CSSimplify.cpp?

Currently, function expressions in `ApplyExpr` are always coerced to rvalues, which is problematic for `mutating func callFunction` method calls. The current workaround in CSApply.cpp is ad-hoc and breaks for `mutating func callFunction` + IUO.

[Here's a small reproducer](https://bugs.swift.org/projects/TF/issues/TF-444), with [`-debug-constraint-solver` output](https://gist.github.com/dan-zheng/378403430d08c5a6471957cfc504bdcf):
```swift

struct Mutating {
  var x: Int
  mutating func callFunction() {
    x += 1
  }
}
func testIUO(f: inout Mutating!) {
  f() // not okay
  f.callFunction() // okay
}
```

```
mutating_iuo.swift:8:3: error: cannot use mutating member on immutable value: 'f' is immutable
  f() // not okay
  ^
```

cc @xedin

rjmccall replied:
The constraint structure of `ApplicableFunction` should be good enough here, so it's probably just a problem in the solver.  Try doing something like this in  `simplifyApplicableFnConstraint` (where the existing similar call to `getFixedTypeRecursive` is):

```
  Type lvType2 = getFixedTypeRecursive(type2, flags, /*wantRValue=*/false);
  type2 = lvType2->hasLValueType() ? getFixedTypeRecursive(lvType2->getRValueType(), flags, /*wantRValue=*/true) : lvType2;
```

Then you can use lvType2 as the base type when building your constraints for the call to `callAsFunction`.

dan-zheng replied:
Thanks for the advice!

I found that the blocker for `mutating func callAsFunction` + IUO. It's the [following line](https://github.com/dan-zheng/swift/blob/fd458d32869960744d0dda8390df0d3bc93efa2d/lib/Sema/CSApply.cpp#L7114) in `ExprRewriter::finishApply` in CSApply.cpp:
```cpp
fn = cs.coerceToRValue(fn);
```

`ConstraintSystem::coerceToRValue` calls `TypeChecker::coerceToRValue`, which calls `cs.setType` on expressions and inserts an implicit load expression. Thus, simply caching `auto *originalFn = fn` before the call to `cs.coerceToRValue(fn)` is not effective.

If I comment the line `fn = cs.coerceToRValue(fn)`, then the `mutating func callAsFunction` + IUO example passes, but other lit tests fail.

Any advice/suggestions? If there's some way to clone `fn` or "undo" the effects of `cs.coerceToRValue(fn)`, that should do the trick. I tried creating an `ASTWalker` to undo the effects of `cs.coerceToRValue(fn)`, but I'm not sure how to nicely undo `ConstraintSystem::setType` which expects a non-`nullptr` `Type` argument.

rjmccall replied:
Can you look at the function expression and/or recorded overload choices for the call and make the decision about callable vs. dynamicCallable vs. function before doing any processing?

dan-zheng replied:
> Can you look at the function expression and/or recorded overload choices for the call and make the decision about callable vs. dynamicCallable vs. function before doing any processing?

Yes! I believe that's possible and may circumvent the issue. Experimenting now.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5632,6 +5632,60 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+    auto &ctx = getASTContext();
+    // Get all call methods of the nominal type.
+    // TODO: Consider caching?
+    SmallVector<FuncDecl *, 4> callMethods;
```
Todo: consider caching the `func call` methods for types?

Such caching of methods is done for [`@dynamicCallable`](https://github.com/apple/swift/blob/master/lib/Sema/ConstraintSystem.h#L1136). There's also a [cache for `@dynamicMemberLookup` types](https://github.com/apple/swift/blob/master/lib/Sema/ConstraintSystem.h#L1139), but it doesn't store methods.

dan-zheng commented:
Verifying that current tests pass.
@swift-ci Please smoke test

lattner commented:
I love how simple and clean this patch is!  Maybe you should mention how simple the implementation is in the proposal?

brentdax commented:
@swift-ci please test compiler performance

dan-zheng committed:
Implement SE-0253 review feedback.

Use `callFunction` instead of `call` as the call-syntax delegate method name.
Add trailing closure tests.

dan-zheng commented:
Updated based on [SE-0253 review feedback](https://forums.swift.org/t/accepted-with-modification-se-0253-callable-values-of-user-defined-nominal-types/24605). The call-syntax delegate method name is now `callFunction` instead of `call`.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented on include/swift/AST/Decl.h:
```diff
@@ -5958,6 +5958,9 @@ class FuncDecl : public AbstractFunctionDecl {
     return getSelfAccessKind() == SelfAccessKind::__Consuming;
   }
+  bool isCallFunction() const {
+    return getName().str() == "callFunction" && isInstanceMember();
```
Todo: unify "call function"/"call method" terminology, and use one spelling consistently.

masters3d replied:
Any plans to diagnose static or class callFunction?

dan-zheng replied:
What diagnostic do you have in mind?

---

masters3d replied:
Precisely. Thanks.

xedin replied:
> A detail is that the current diagnostics are being ported from the old system (CSDiag.cpp) to the new request-based system, so adding diagnostics to the old system might be throwaway work. 

You can actually add diagnostics to a new diagnostic framework right now, it's live, in fact me and couple of others have ported a lot of diagnostics already, see https://github.com/apple/swift/pull/26074 for the most recent example of how easy it is to add new diagnostics now.

rjmccall reviewed (changes requested):
Per the discussion in the review announcement thread, please rename the key function here `callAsFunction`.

dan-zheng committed:
Merge branch 'master' into callables

dan-zheng committed:
Implement SE-0253 review feedback.

Use `callAsFunction` as the call-syntax delegate method name.

dan-zheng commented:
> Per the discussion in the review announcement thread, please rename the key function here `callAsFunction`.

Done in https://github.com/apple/swift/pull/24299/commits/518bda2c5d29bca905d2e4911bb60bf2e1a2e210.

Some notes, in order of importance:
* `mutating func callAsFunction` doesn't work with IUO. See [this comment](https://github.com/apple/swift/pull/24299/files#r285090012) for more details. I couldn't think of better solutions and need help from constraint system experts.
* Diagnostics for sugared call-syntax calls have room for improvement. This can be done later. Example of current diagnostics:

  ```swift
  struct SimpleCallable {
    func callAsFunction(_ x: Float) -> Float {
      return x
    }
  }
  
  let foo = SimpleCallable()
  
  _ = foo.callAsFunction(1, 1)
  // error: extra argument in call
  // _ = foo.callAsFunction(1, 1)
  //                           ^
  
  _ = foo(1, 1)
  // error: cannot call value of non-function type 'SimpleCallable'
  // _ = foo(1, 1)
  //     ~~~^
  // error: cannot invoke 'foo' with an argument list of type '(Int, Int)'
  // _ = foo(1, 1)
  //     ^
  ```

* Should `func callAsFunction` methods be cached for nominal types? I haven't done so yet (grep `Note: consider caching` for places that benefit from caching), but I imagine it would help type-checker performance for code that heavily uses call-syntax functionality.

brentdax commented on include/swift/AST/Decl.h:
```diff
@@ -5958,6 +5958,9 @@ class FuncDecl : public AbstractFunctionDecl {
     return getSelfAccessKind() == SelfAccessKind::__Consuming;
   }
+  bool isCallAsFunctionMethod() const {
+    return getName().str() == "callAsFunction" && isInstanceMember();
```
It'd be better to test against `getASTContext().Id_callAsFunction` here.

dan-zheng replied:
That would be nice. Unfortunately, I get this error:
```
/Users/danielzheng/swift-master/swift/include/swift/AST/Decl.h:5962:46: error: member access into incomplete type 'swift::ASTContext'
    return getName().str() == getASTContext().Id_callAsFunction &&
                                             ^
/Users/danielzheng/swift-master/swift/include/swift/AST/Identifier.h:31:9: note: forward declaration of 'swift::ASTContext'
  class ASTContext;
        ^
```

brentdax replied:
Right, you'll need to make this a forward declaration and move the body into Decl.cpp, sorry. (And also drop the `.str()` on the left side of the `==`.)

dan-zheng replied:
Ah right, thanks! Done in https://github.com/apple/swift/pull/24299/commits/9695a26dc7c4f7b32dcf48d81d2c5b7b93496717.

dan-zheng committed:
Address review feedback.

rjmccall commented on lib/Sema/CSApply.cpp:
```diff
@@ -7197,6 +7200,59 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+    apply->setArg(arg);
+    cs.setType(apply, methodType->getResult());
+    cs.cacheExprTypes(apply);
+    return apply;
```
Can you extract everything in here that's not part of asking whether the call is to a `callAsFunction` method into a separate function?  You can pass the overload choice down to avoid having to recompute it.  (That might also be useful for the dynamic-callable path.)

dan-zheng replied:
Code extracted to `finishApplyCallAsFunctionMethod` in https://github.com/apple/swift/commit/fd458d32869960744d0dda8390df0d3bc93efa2d.

I didn't pass the overload choice because I'm not exactly sure which calls should be optimized to avoid recomputation.

rjmccall commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5833,6 +5833,60 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+      addConstraint(ConstraintKind::ApplicableFunction, funcType, tv, locator);
+
+      return SolutionKind::Solved;
+    }
```
Same thing: please extract this into a helper function; you can have it return whether it succeeded.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24299/commits/fd458d32869960744d0dda8390df0d3bc93efa2d.

dan-zheng commented on include/swift/AST/Decl.h:
```diff
@@ -5958,6 +5958,7 @@ class FuncDecl : public AbstractFunctionDecl {
   bool isConsuming() const {
     return getSelfAccessKind() == SelfAccessKind::__Consuming;
   }
+  bool isCallAsFunctionMethod() const;
```
Done in https://github.com/apple/swift/pull/24299/commits/fd458d32869960744d0dda8390df0d3bc93efa2d.

rjmccall commented:
Thanks.  I haven't given the constraint logic a good once-over; do you want to get `mutating` calls working first?

dan-zheng commented:
> Thanks. I haven't given the constraint logic a good once-over; do you want to get `mutating` calls working first?

Sure thing. I missed your [earlier suggestion](https://github.com/apple/swift/pull/24299#discussion_r288878690) regarding `mutating` calls, I'll explore that and get back to you.

rxwei commented:
Would it be acceptable merge this partial implementation as is and start separate PRs later to support `mutating` and IUO?

dan-zheng committed:
Merge remote-tracking branch 'apple/master' into callables

dan-zheng commented:
> Would it be acceptable merge this partial implementation as is and start separate PRs later to support `mutating` and IUO?

I overlooked [this suggestion](https://github.com/apple/swift/pull/24299/files#r289252065) from @rjmccall which might resolve the `mutating` + IUO issue. I'll investigate and report back today.

dan-zheng committed:
Fix `mutating func callAsFunction` + IUO.

- Reorganize `ExprRewriter::finishApply` to check `callAsFunction`
  and `@dynamicCallable` cases before coercing the function expression
  to an rvalue.
- Enable `mutating callAsFunction` + IUO test.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test

rjmccall commented:
This looks good to me, but maybe @xedin should give the CS logic a new once-over since it's changed somewhat.

dan-zheng commented:
@swift-ci Please test compiler performance

dan-zheng commented:
@swift-ci Please test compiler performance

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5969,6 +5969,60 @@ ConstraintSystem::simplifyApplicableFnConstraint(
   }
 
+  // Handle applications of types with `callAsFunction` methods.
+  if (desugar2->mayHaveMembers()) {
```
In addition to that it would be great to add a new path element for this kind of situation, something like `implicit call` so it could easy easily identified and diagnosed.

dan-zheng replied:
Thanks for your suggestions @xedin! I explored your suggestions and got stuck (`mutating func callAsFunction` doesn't work with member constraint) - posted info in [this comment](https://github.com/dan-zheng/swift/commit/394c589198288f41c64012e182bc85faa530354c#commitcomment-34646601). Could you please take a look? 🙂

xedin replied:
Happy to help! I have replied on that comment, it should be pretty easy to fix.

dan-zheng replied:
`func callAsFunction` constraint revamp done in https://github.com/apple/swift/pull/24299/commits/8dcd1c68b2acbf6fffc664b853b168a851584ea5 - please read the commit description for more details.

@xedin: could you please re-review?

xedin commented:
Regarding diagnostics it would be great to know what the plan is here, since it seems like it's always secondary concern. It would be great if you could add something to proposal to make it easier for others to pick that work up.

dan-zheng commented on lib/Sema/CSApply.cpp:
```diff
@@ -6792,11 +6792,70 @@ Expr *ExprRewriter::convertLiteralInPlace(Expr *literal,
+                              diag::cannot_pass_rvalue_mutating_subelement,
+                              diag::cannot_pass_rvalue_mutating);
+    failure.diagnose();
+    return nullptr;
```
Yes, this is definitely suboptimal. Could you please give suggestions on avoiding this fallback constraint in CSApply, e.g. adding some constraints in CSSimplify?

xedin replied:
I think building member constraint with new simplify constraint should solve that problem because everything would go through primary code path.

dan-zheng replied:
Constraint revamp done in https://github.com/apple/swift/commit/8dcd1c68b2acbf6fffc664b853b168a851584ea5.

In CSSimplify: manual overload set construction for `func callAsFunction` has been replaced with a member constraint. This made it possible to remove this CSApply hack for diagnosing invalid `mutating func callAsFunction` applications, though the [diagnostics are suboptimal](https://github.com/apple/swift/pull/24299/files/35607ed25aae5a09c782259e1960513974c6a964..8dcd1c68b2acbf6fffc664b853b168a851584ea5#r312724239).

xedin commented:
As a follow-up to my previous question, I have also wrote up thoughts I have regarding improving process around diagnostics  and reasons why I think it's important [on Discussion forum](https://forums.swift.org/t/additions-to-proposal-process-to-document-feature-syntax-use-and-error-warning-scenarios/26037)

dan-zheng commented:
Thanks @xedin! I'm unfortunately caught up in other work, but I'll try to revisit this PR later this month and address your comments.

---

> Regarding diagnostics it would be great to know what the plan is here, since it seems like it's always secondary concern. It would be great if you could add something to proposal to make it easier for others to pick that work up.

Good diagnostics are definitely important. I think a good goal for `func callAsFunction` is for sugared applications to produce the same diagnostics as direct member applications, which currently isn't the case. I marked test files with `TODO: Improve this error...` comments:

```swift
// TODO: Improve this error to match the error using a direct `callAsFunction` member reference.
// expected-error @+2 {{cannot call value of non-function type 'SimpleCallable'}}
// expected-error @+1 {{cannot invoke 'foo' with an argument list of type '(Int, Int)'}}
_ = foo(1, 1)
// expected-error @+1 {{cannot convert value of type 'SimpleCallable' to specified type '(Float) -> Float'}}
let _: (Float) -> Float = foo
```

After this PR lands, I'll file a bug for improving diagnostics so that others can help.

xedin commented:
@dan-zheng I think if you use member constraint, you'd get some of the diagnostics for free since a lot has been ported to new diagnostic framework already.

dan-zheng committed:
Merge remote-tracking branch 'apple/master' into callables

dan-zheng committed:
Clean up CSSimplify and CSApply for `func callAsFunction`.

- CSSimplify: get type variable for `func callAsFunction` member by
  generating a member constraint (with the potentially-lvalue base `type2`)
  rather than manually constructing an overload set.
- CSApply: remove hack for diagnosing invalid `mutating callAsFunction` calls.
- Add todo comments for improving
  `ConstraintSystem::simplifyDynamicCallableApplicableFnConstraint`.

dan-zheng committed:
Add Optional `callAsFunction` test.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+                     return isa<FuncDecl>(entry.getValueDecl());
+                   });
+  if (hasCallAsFunctionMethods) {
+    auto memberLoc = getConstraintLocator(
```
I'm not sure that usages of `memberLoc` or `locator` are correct/ideal here.

@xedin: could you please check these `ConstraintLocator *` usages?
You suggested [adding an "implicit call" path element](https://github.com/apple/swift/pull/24299#discussion_r298708877), but I'm not sure where to use it - please advise!

xedin replied:
I think for `addValueMemberConstraint` it makes more sense to use `memberLoc` rather than `locator`, and for `ApplicableFunction` indeed `locator` is best.

Regarding `implicit call` you don't have to worry about it right now, we can do it in the follow-up instead. I was thinking that it would be great to associate such new element with locator which belongs to `ApplicableFunction` something like `<call anchor> -> implicit call -> apply function` so when argument conversions are expanded by `matchCallArguments` they'd get `<call anchor> -> implicit call -> apply arg-to-param -> comparing #N to #M` locators.

dan-zheng commented on test/Sema/call_as_function_simple.swift:
```diff
@@ -106,8 +106,12 @@ struct Mutating {
 func testMutating(_ x: Mutating, _ y: inout Mutating) {
-  _ = x() // expected-error {{cannot use mutating member on immutable value: 'x' is a 'let' constant}}
-  _ = x.callAsFunction() // expected-error {{cannot use mutating member on immutable value: 'x' is a 'let' constant}}
+  // TODO: Improve this error to match the error using a direct `callAsFunction` member reference.
```
Note: this diagnostic regressed: the diagnostic is now produced from failed `member` constraint instead of custom `CSApply` diagnostic hack. cc @xedin regarding ideas for improving/fixing this.

The diagnostic below in `testInout` has improved however - it is now in parity with direct `callAsFunction` calls.

dan-zheng commented:
I believe the main constraint system changes requested by @xedin are done, so this patch is ready for re-review. Some `func callAsFunction` diagnostics are suboptimal though - details in comments above.

Verifying that tests pass.
@swift-ci Please test

dan-zheng commented:
Verifying that tests pass.
@swift-ci Please test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+    // type variable.
+    auto memberTy = createTypeVariable(memberLoc, TVO_CanBindToLValue |
+                                                  TVO_CanBindToNoEscape |
+                                                  TVO_CanBindToInOut);
```
I don't think member type could be bound to any of these types listed in flags btw since it represents a result type of the member choice, `inout` makes sense only to arguments same way as `l-value` and can't really return no-escape function right?

dan-zheng replied:
I see, that makes sense. How can I call `createTypeVariable` with no flag arguments?

```
swift/lib/Sema/CSSimplify.cpp:6108:49: error: too few arguments to function call, expected 2, have 1
    auto memberTy = createTypeVariable(memberLoc);
                    ~~~~~~~~~~~~~~~~~~          ^
```

xedin replied:
`createTypeVariable(memberLoc, /*options=*/0);

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24299/commits/cebbd6335dab7e0bbf3e5c0dff9c445588e01f81.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+                                                  TVO_CanBindToInOut);
+    addValueMemberConstraint(origType2, DeclName(ctx.Id_callAsFunction),
+                             memberTy, DC, FunctionRefKind::SingleApply,
+                             /*outerAlternatives*/ {}, locator);
```
Might make sense to add a `TODO` in case we'd ever want to get a static variant of `callAsFunction`...

dan-zheng replied:
I see - would you like the `TODO` comment to say anything more than ``TODO: Revisit this if `static func callAsFunction` is to be supported.``?

I'm also curious about what constraint changes are necessary to support `static func callAsFunction`.

xedin replied:
This is about `SingleApply` vs. `DoubleApply` I believe `originalTy2` would already be a metatype value so nothing would have to be done there.

dan-zheng replied:
I see, that makes sense. (I guess static functions need `DoubleApply` due to metatype currying.)

dan-zheng replied:
Todo hint added in https://github.com/apple/swift/pull/24299/commits/cebbd6335dab7e0bbf3e5c0dff9c445588e01f81.

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -6748,11 +6751,62 @@ Expr *ExprRewriter::convertLiteralInPlace(Expr *literal,
+// Returns true if the given method and method type are a valid
+// `@dynamicCallable` required `func dynamicallyCall` method.
+static bool isValidDynamicallyCallMethod(FuncDecl *method,
+                                         AnyFunctionType *methodType) {
```
Maybe a better name would be `isValidForDynamicCall` or similar?

dan-zheng replied:
I don't believe "dynamic call" is a term of art, so I would prefer a name that involves "dynamicCallable" or "dynamicallyCallMethod".

How about `isValidDynamicCallableMethod`?

xedin replied:
Sure, that works too.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24299/commits/cebbd6335dab7e0bbf3e5c0dff9c445588e01f81.

xedin reviewed (changes requested):
Looks much better now, thank you! Still needs a final push though.

dan-zheng commented:
There's one failing test (`expr/delayed-ident/static_var.swift)` as of the current constraint system changes in https://github.com/apple/swift/pull/24299/commits/8dcd1c68b2acbf6fffc664b853b168a851584ea5:
```
Assertion failed: (unwrappedType && "Unwrapped type must not be null"), function ForceOptional, file /Users/danielzheng/swift-master/swift/lib/Sema/CSFix.h, line 272.
Stack dump:
0.	Program arguments: /Users/danielzheng/swift-master/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/bin/swiftc -frontend -target x86_64-apple-macosx10.9 -module-cache-path /Users/danielzheng/swift-master/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -swift-version 4 -typo-correction-limit 10 -typecheck -verify -disable-objc-attr-requires-foundation-module /Users/danielzheng/swift-master/swift/test/expr/delayed-ident/static_var.swift
1.	Swift version 5.1-dev (LLVM 200186e28b, Swift 4cb7ebf8a0)
2.	While type-checking statement at [/Users/danielzheng/swift-master/swift/test/expr/delayed-ident/static_var.swift:49:1 - line:49:34] RangeText="var _: HasClosure = .factoryOpt(3"
3.	While type-checking declaration 0x7fc815817b18 (at /Users/danielzheng/swift-master/swift/test/expr/delayed-ident/static_var.swift:49:1)
4.	While type-checking expression at [/Users/danielzheng/swift-master/swift/test/expr/delayed-ident/static_var.swift:49:21 - line:49:34] RangeText=".factoryOpt(3"
0  swiftc                   0x000000010d58f215 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swiftc                   0x000000010d58e258 llvm::sys::RunSignalHandlers() + 248
2  swiftc                   0x000000010d58f808 SignalHandler(int) + 264
3  libsystem_platform.dylib 0x00007fff6aeeeb5d _sigtramp + 29
4  swiftc                   0x000000010f96c2c3 cmark_strbuf__initbuf + 128062
5  libsystem_c.dylib        0x00007fff6ada86a6 abort + 127
6  libsystem_c.dylib        0x00007fff6ad7120d basename_r + 0
7  swiftc                   0x000000010d89c003 swift::constraints::ForceOptional::create(swift::constraints::ConstraintSystem&, swift::Type, swift::Type, swift::constraints::ConstraintLocator*) (.cold.2) + 35
8  swiftc                   0x000000010a2d282c swift::constraints::ForceOptional::create(swift::constraints::ConstraintSystem&, swift::Type, swift::Type, swift::constraints::ConstraintLocator*) + 108
9  swiftc                   0x000000010a2ae96f swift::constraints::ConstraintSystem::simplifyApplicableFnConstraint(swift::Type, swift::Type, swift::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) + 3007
10 swiftc                   0x000000010a2b2cbc swift::constraints::ConstraintSystem::simplifyConstraint(swift::constraints::Constraint const&) + 1980
11 swiftc                   0x000000010a2bb8c7 swift::constraints::ConstraintSystem::simplify(bool) + 87
12 swiftc                   0x000000010a26121c swift::constraints::TypeVariableBinding::attempt(swift::constraints::ConstraintSystem&) const + 412
13 swiftc                   0x000000010a2cc7a4 swift::constraints::BindingStep<swift::constraints::TypeVarBindingProducer>::take(bool) + 468
14 swiftc                   0x000000010a2bd162 swift::constraints::ConstraintSystem::solve(llvm::SmallVectorImpl<swift::constraints::Solution>&) + 338
15 swiftc                   0x000000010a3144dd swift::constraints::ConstraintSystem::salvage(llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::Expr*) + 285
16 swiftc                   0x000000010a2be89c swift::constraints::ConstraintSystem::solve(swift::Expr*&, swift::Type, swift::ExprTypeCheckListener*, llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::FreeTypeVariableBinding) + 508
```

I'm not exactly sure what is the cause.
@xedin: is it possible that this change is related to the change in definition of [`origType2`](https://github.com/apple/swift/pull/24299/commits/8dcd1c68b2acbf6fffc664b853b168a851584ea5#diff-69306f7b0822cf52e8c0639307ad7407L6086)?

xedin commented:
Maybe, I'm not sure either... Maybe `ForceOptional` has to get l-value stripped away before giving type to a fix?

dan-zheng committed:
Address review comments, fix test.

- Rename `isValidDynamicallyCallMethod` to `isValidDynamicCallableMethod`.
- Add todo hint regarding `static func callAsFunction`, if it is to be supported.
- Restore `var origType2` before unwrapping optionality.
  - This fixes `expr/delayed-ident/static_var.swift`.

dan-zheng commented:
I believe all comments have been addressed (besides improving diagnostics).
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test macOS

xedin reviewed (approved):
LGTM! Thank you!

dan-zheng committed:
Reference SR-11378 in todo comments: improve `callAsFunction` diagnostics.
